cpp/util/thread_pool.cc

```diff
@@ -56,30 +78,40 @@ ThreadPool::Impl::~Impl() {
 
 void ThreadPool::Impl::Worker() {
   while (true) {
-    function<void()> closure;
+    QueueEntry entry;
 
     {
       unique_lock<mutex> lock(queue_lock_);
-
-      // If there's nothing to do, wait until there is.
-      if (queue_.empty()) {
-        queue_cond_var_.wait(lock);
-
-        // condition_variable::wait can return spuriously.
-        if (queue_.empty())
-          continue;
+      while (queue_.empty() || get<0>(queue_.top()) > system_clock::now()) {
+        const system_clock::duration duration(
+            queue_.empty() ? seconds::max()
+                           : get<0>(queue_.top()) - system_clock::now());
+        // If there's nothing to do, wait until there is.
+        queue_cond_var_.wait_for(lock, duration);
       }
 
-      // If we received an empty closure, exit cleanly.
-      if (!queue_.front())
-        break;
-
-      closure = queue_.front();
+      entry = queue_.top();
       queue_.pop();
+
+      // If we received an empty entry, exit cleanly.
+      if (!get<1>(entry)) {
+        // Anything left in the queue must be either other exit sentintels, or
+        // future (delayed) tasks, so we'll cancel anything we find, up until
+        // either the next exit sentinel, or the end of the queue.
+        VLOG(1) << "Cancelling delayed tasks...";
+        int cancelled(0);
+        while (!queue_.empty() && get<2>(queue_.top())) {
+          get<2>(queue_.top())->Return(util::Status::CANCELLED);
```

Comment:
Actually, what'll happen at the moment is that the `Add()` will block forever because the cancelling is done under `queue_lock`, I can sort that and then you'd hit the CHECK in the d'tor.